contracts/infrastructure/DappRegistry.sol

```diff
@@ -2,77 +2,197 @@ pragma solidity ^0.6.12;
+    event RegistryCreated(uint8 registryId, address registryOwner);
+    event RegistryRemoved(uint8 registryId);
+    event OwnerChanged(uint8 registryId, address newRegistryOwner);
+    event TimelockChangeRequested(uint64 newSecurityPeriod);
+    event TimelockChanged(uint64 newSecurityPeriod);
+    event FilterUpdated(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event FilterUpdateRequested(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event DappAdded(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event DappRemoved(uint8 indexed registryId, address dapp);
 
-    bytes32[] public registries;
+    modifier onlyOwner(uint8 _registryId) {
+        validateOwner(_registryId);
+        _;
+    }
+    
+    constructor(uint64 _securityPeriod) public {
+        // set the timelock period
+        securityPeriod = _securityPeriod;
+        // set the owner of the Argent Registry (registryId == 0)
+        registryOwners[0] = msg.sender;
+    }
 
-    mapping (bytes32 => address) public registryManager;
+    /********* Wallet-centered functions *************/
 
-    event RegistryCreated(bytes32 registry, address manager);
-    event RegistryRemoved(bytes32 registry);
+    function isEnabledRegistry(address _wallet, uint8 _registryId) external view returns (bool) {
+        uint registries = uint(enabledRegistryIds[_wallet]);
+        return ((registries >> _registryId) & 1) > 0;
+    }
 
-    function authorise(address _wallet, address _contract, bytes calldata _data) external view override returns (bool) {
-        (bool isActive, address filter) = getFilter(_wallet, _contract);
+    function isAuthorised(address _wallet, address _spender, address _to, bytes calldata _data) external view override returns (bool) {
+        (bool isActive, address filter) = getFilter(_wallet, _spender);
         if (isActive) {
-            return _data.length == 0 || filter == address(0) || IFilter(filter).validate(_data);
+            return _data.length == 0 || filter == address(0) || IFilter(filter).validate(_wallet, _spender, _to, _data);
         }
         return false;
     }
 
-    function toggle(address _wallet, bytes32 _registry) external override onlyModule(_wallet) returns (bool) {
-        require(_registry == ARGENT_DEFAULT || registryManager[_registry] != address(0), "AR: unknow registry");
-        bool current = isEnabled[_wallet][_registry];
-        isEnabled[_wallet][_registry] = !current;
-        return !current;
+    function toggleRegistry(address _wallet, uint8 _registryId, bool _enabled) external override onlyModule(_wallet) returns (bool) {
+        require(registryOwners[_registryId] != address(0), "AR: unknown registry");
+        uint registries = uint(enabledRegistryIds[_wallet]);
+        bool current = ((registries >> _registryId) & 1) > 0;
+        require(current != _enabled, "AR: bad state change" );
```

Comment:
We can do without these two lines and save ~2,000 gas. I can't see any harm that could be done with someone trying to toggle a registry setting to an existing value. The `^` would just result in the same value being passed in `enabledRegistryIds[_wallet]`. The cost of toggling between differencing values is ~47,000 and to an existing value is ~15,000 gas.

Reply:
This was introduced to meet a requirement from @th3m477 of having explicit enable/disable calls instead of allowing the same function call to do both. I assume this is to mitigate the risk of having a client mistakenly sign the same toggle transaction twice? I think this was discussed last Monday when I was off so I'll let Matt explain the his reasons :)

Reply:
Had a conversation with Matt and Julien on the call and decided we can remove this logic.

Reply:
@olivdb it's ok to not have the check 👍 

i just wanted `_enabled` passed in which i can see you do, so I am happy

Reply:
I changed the `require` into an `if`

Reply:
To save all the gas we can get rid of that `if`. I don't see the point of that check. If the value has no change, then no storage will be written.

Reply:
Without the `if`, the XOR will flip the bit no matter what, making it inconsistent with the value of `_enabled` when `current == _enabled`. An alternative to the if is to set/unset the bit based on `_enabled`
```
        if(_enabled) {
            enabledRegistryIds[msg.sender] = bytes32(registries | (uint(1) << _registryId));
        } else {
            enabledRegistryIds[msg.sender] = bytes32(registries & ~(uint(1) << _registryId));
        }
  ```
  Same gas cost but maybe a bit more verbose.